opm/input/eclipse/Schedule/ResCoup/GrupSlav.cpp

```diff
@@ -0,0 +1,157 @@
+#include <opm/common/utility/OpmInputError.hpp>
+#include "../HandlerContext.hpp"
+
+#include <fmt/format.h>
+
+#include <limits>
+#include <stdexcept>
+
+namespace Opm {
+namespace ReservoirCoupling {
+GrupSlav GrupSlav::serializationTestObject()
+{
+    return GrupSlav{"MANI-D", "D1-M", FilterFlag::MAST, FilterFlag::MAST, FilterFlag::MAST,
+                    FilterFlag::MAST, FilterFlag::MAST, FilterFlag::MAST, FilterFlag::MAST};
+}
+
+bool GrupSlav::operator==(const GrupSlav& rhs) const {
+    return
+        this->m_name == rhs.m_name &&
+        this->m_master_group_name == rhs.m_master_group_name &&
+        this->m_oil_prod_flag == rhs.m_oil_prod_flag &&
+        this->m_liquid_prod_flag == rhs.m_liquid_prod_flag &&
+        this->m_gas_prod_flag == rhs.m_gas_prod_flag &&
+        this->m_fluid_volume_prod_flag == rhs.m_fluid_volume_prod_flag &&
+        this->m_oil_inj_flag == rhs.m_oil_inj_flag &&
+        this->m_water_inj_flag == rhs.m_water_inj_flag &&
+        this->m_gas_inj_flag == rhs.m_gas_inj_flag;
+}
+
+// NOTE: This function is needed by Boost.Test, see test/ReservoirCoupling/GrupSlavTest.cpp
+std::ostream& operator<<(std::ostream& os, const GrupSlav::FilterFlag& flag) {
+    switch (flag) {
+        case GrupSlav::FilterFlag::MAST:
+            os << "MAST";
+            break;
+        case GrupSlav::FilterFlag::SLAV:
+            os << "SLAV";
+            break;
+        case GrupSlav::FilterFlag::BOTH:
+            os << "BOTH";
+            break;
+        default:
+            throw std::invalid_argument("Invalid filter flag");
+    }
+    return os;
+}
+
+} // namespace ReservoirCoupling
+
+void checkValidSlaveGroupName(const std::string& name, HandlerContext& handlerContext)
+{
+    auto rescoup = handlerContext.state().rescoup();
+    if (rescoup.hasGrupSlav(name)) {
+        std::string msg = fmt::format("GRUPSLAV group {} already defined. Redefining", name);
+        OpmLog::warning(OpmInputError::format(msg, handlerContext.keyword.location()));
+    }
+    auto groups = handlerContext.state().groups;
+    if (!groups.has(name)) {
+        std::string msg = fmt::format("Group '{}': Not defined. Slave groups should be defined in advance by using GRUPTREE or WELSPECS before referenced in GRUPSLAV.", name);
+        throw OpmInputError(msg, handlerContext.keyword.location());
+    }
+}
+
+ReservoirCoupling::GrupSlav::FilterFlag getFilterFlag(const DeckItem& keyword, HandlerContext& handlerContext)
+{
+    try {
+        return ReservoirCoupling::GrupSlav::filterFlagFromString(keyword.getTrimmedString(0));
+    } catch (const std::invalid_argument& e) {
+        std::string msg = fmt::format("Invalid filter flag: {}", keyword.getTrimmedString(0));
+        throw OpmInputError(msg, handlerContext.keyword.location());
+    }
+}
+
+void handleGRUPSLAV(HandlerContext& handlerContext)
+{
+    auto schedule_state = handlerContext.state();
+    auto rescoup = schedule_state.rescoup();
```

Comment:
IS there a reason for the copy or would it be better to use `const auto&`?

Reply:
Here we should probably make a copy of `rescoup` object, since we later (on line 149) update the object (so it cannot be const) and on line 154 we update the schedule state with the object, so we do not want to modify the previous schedule state objects and we should make a copy.